python/TestHarness/TestHarness.py

```diff
@@ -235,85 +240,145 @@ def launchTest(self, find_only, file, dirpath, testers):
+                            command = self.createQsubFile(command, dirpath, tester)
+                    # This method spawns another process and allows this loop to continue looking for tests
+                    # RunParallel will call self.testOutputAndFinish when the test has completed running
+                    # This method will block when the maximum allowed parallel processes are running
+                    self.runner.run(tester, command)
+                else: # This job is skipped - notify the runner
+                    status = tester.getStatus()
+                    if status != tester.bucket_silent: # SILENT occurs when a user is using --re options
+                        if (self.options.report_skipped and status == tester.bucket_skip) \
+                           or status == tester.bucket_skip:
+                            self.handleTestStatus(tester)
+                        elif status == tester.bucket_deleted and self.options.extra_info:
+                            self.handleTestStatus(tester)
+                    self.runner.jobSkipped(tester.parameters()['test_name'])
+
+    # Create the PBS qsub script, perform directory/file operations
+    def createQsubFile(self, command, dirpath, tester):
+        pbs_job_dir = os.path.join(dirpath, 'job_' + self.options.pbs)
+        # Modify the tester's test_dir to point to the new location we just created
+        tester.specs['test_dir'] = pbs_job_dir
+
+        if not os.path.exists(pbs_job_dir):
+            os.mkdir(pbs_job_dir)
+        os.chdir(pbs_job_dir)
+
+        pbs_params = self.factory.validParams('PBSJob')
+        job_instance = self.factory.create('PBSJob', tester.specs['test_name'], pbs_params)
+        specs = job_instance.specs
+
+        ## Set some PBS specific specs that help populate the qsub script (and directory)
+        specs['job_name'] = tester.specs['pbs_job_name']
+        specs['template_script'] = self.moose_dir + '/python/TestHarness/pbs_submit.sh'
+        specs['no_copy_pattern'] = '.*\.sh'
+        specs['mpi_procs'] = tester.getProcs(self.options)
+        specs['soft_link_output'] = False
+        specs['place'] = 'free'
+        specs['combine_streams'] = '#PBS -j oe'
+
+
+        #### create a set for copy and nocopy so its easier to work with
+        pbs_nocopy_files = set([self.options.input_file_name])
+        pbs_copy_files = set([])
+
+        # Exlude any additional files
+        if tester.specs.isValid('pbs_nocopy_files'):
+            pbs_nocopy_files.update(tester.specs['pbs_nocopy_files'])
+
+        # Include any additional files needed for the test
+        if tester.specs.isValid('pbs_copy_files'):
+            pbs_copy_files.update(tester.specs['pbs_copy_files'])
+
+        # Some tests specify a different gold directory
+        if tester.specs.isValid('gold_dir'):
+            pbs_copy_files.update([tester.specs['gold_dir']])
+
```

Comment:
I think I am using 'isValid' incorrectly. I found every one of these logic statements was always True during testing.

Because we are working with sets, we could just update the set with an potentially empty list instead... or instead of using inValid, perform some len() operation?